### ISSUE TITLE
xpra coredumping - selinux?

### ISSUE THREAD

**Issue by dwbotsch** [Xpra-org/xpra#3243]
On our RHEL8 boxes, it looks like xpra recently started a coredump loop. I'm guessing that part of the problem is selinux. Looking at the output of ausearch, I'm wondering why xpra is trying to access all this stuff and if needed, why that isn't part of the included selinux rules?
Does xpra really need all this?
```
ausearch -c 'xpra' --raw | audit2allow


#============= xpra_t ==============
allow xpra_t admin_home_t:dir write;
allow xpra_t binfmt_misc_fs_t:dir write;
allow xpra_t device_t:dir write;
allow xpra_t home_root_t:dir write;
allow xpra_t hugetlbfs_t:dir write;
allow xpra_t nfs_t:dir read;
allow xpra_t security_t:dir write;
allow xpra_t self:capability sys_resource;

#!!!! This avc can be allowed using the boolean 'domain_can_mmap_files'
allow xpra_t tmp_t:file map;
allow xpra_t tmpfs_t:file execute;
allow xpra_t user_tmp_t:dir write;

#!!!! This avc can be allowed using the boolean 'domain_can_mmap_files'
allow xpra_t xpra_pid_t:file map;
```

**Comment by totaam** [comment]
> I'm guessing that part of the problem is selinux

You could try running the server by hand with `-d all` and see what the last messages in the log files say. Or run in gdb to get a meaningful backtrace.

I'm going to assume that this is RHEL 8.4.
Perhaps they've changed something in the policy in 8.4, or perhaps a path change in xpra is now triggering other rules. It could also be a bug in xpra.
You could just run with selinux disable to see if that is the cause of the "coredump loop".
As for which rules are really needed, that's impossible to say from the information above as the paths are not included.

You've also omitted the full version details and the command lines used.
At least the mmap part seems legit. This is needed if you intend to use mmap.
You can avoid these warnings by running with `--no-mmap`.

**Comment by dwbotsch** [comment]
Yup... RHEL8.4 .

rpm -q xpra
xpra-4.2.2-10.r29635.el8.x86_64

and systemctl start xpra .

Haven't had a chance to troubleshoot further as suggested, yet.
thanks.

**Comment by totaam** [comment]
It's such a pain to debug because you can't run this interactively or it doesn't crash..
The problem seems to come from the `SIGCHLD` handler, the only thing we do there is that we re-schedule via `GLib.idle_add` to avoid doing anything from the signal handler thread.
But it seems that the gi bindings are unable to do that and spew out`could not allocate closure` before crashing.
```
Traceback (most recent call first):
  File "/usr/lib64/python3.6/site-packages/gi/overrides/GLib.py", line 601, in idle_add
    return GLib.idle_add(priority, function, *user_data)
  File "/usr/lib64/python3.6/site-packages/xpra/child_reaper.py", line 156, in sigchld
    GLib.idle_add(self._sigchld, signum, str(frame))
  File "/usr/lib64/python3.6/warnings.py", line 398, in __init__
    def __init__(self, message, category, filename, lineno, file=None,
  File "/usr/lib64/python3.6/site-packages/gi/overrides/GLib.py", line 498, in run
    super(MainLoop, self).run()
  File "/usr/lib64/python3.6/site-packages/xpra/server/proxy/proxy_server.py", line 196, in do_run
    self.main_loop.run()
  File "/usr/lib64/python3.6/site-packages/xpra/server/server_core.py", line 422, in run
    self.do_run()
  File "/usr/lib64/python3.6/site-packages/xpra/scripts/server.py", line 1312, in do_run_server
    log.info("%s", m)
  File "/usr/lib64/python3.6/site-packages/xpra/scripts/main.py", line 1835, in run_server
    return do_run_server(script_file, cmdline, error_cb, options, args, mode, display, defaults)
  File "/usr/lib64/python3.6/site-packages/xpra/scripts/main.py", line 442, in do_run_mode
    return run_server(script_file, cmdline, error_cb, options, args, mode, defaults)
  File "/usr/lib64/python3.6/site-packages/xpra/scripts/main.py", line 412, in run_mode
    return do_run_mode(script_file, cmdline, error_cb, options, args, mode, defaults)
  File "/usr/lib64/python3.6/site-packages/xpra/scripts/main.py", line 139, in main
    return run_mode(script_file, cmdline, err, options, args, mode, defaults)
  File "/usr/bin/xpra", line 10, in <module>
    sys.exit(main(sys.argv[0], sys.argv))
```

Seems similar to: [RHEL8.2 setroubleshootd crashes when ffi closure cannot execute](https://bugzilla.redhat.com/show_bug.cgi?format=multiple&id=1871869)
Which seems to be related to the ability to mmap something, or permissions on `/tmp` - not sure which.

We even have a note in the xpra source code about how we need to register `SIGCHLD` before calling `get_platform_info` because the call will spawn a subprocess which exits and triggers the signal.
In any case, even removing the call doesn't fix things as the crash still occurs, only later: it seems impossible to run the glib main loop in this broken state..

I think I have wasted enough time on this bug. Perhaps RedHat can fix it?
Until then "disable SELinux" may be the "solution".

**Comment by totaam** [comment]
_Please note that the fact that libffi cannot execute when all temporary file systems are marked with "nodev" has been fixed through [BZ #1875340](https://bugzilla.redhat.com/show_bug.cgi?id=1875340) (RHEL 8.6.0), so now it's usually expected to run fine._

_libffi upstream has fixed this by offering a build options for static closures which do not depend at all on run-time code generation anymore. This work is not really backportable._

This upstream problem can be resolved by updating your OS.
